hoomd/hpmc/ShapeMoves.h

```diff
@@ -0,0 +1,870 @@
+        Scalar current_volume = mp.getVolume();
+        vec3<Scalar> dr = this->m_centroids[typid] - mp.getCenterOfMass();
+        OverlapReal scale = (OverlapReal)fast::pow(volume / current_volume, 1.0 / 3.0);
+        this->scaleParticleVolume(shape, dr, scale);
+        this->m_mc->setParam(typid, shape);
+        }
+
+    void scaleParticleVolume(param_type& shape, vec3<Scalar> dr, OverlapReal scale)
+        {
+        Scalar rsq = 0.0;
+        for (unsigned int i = 0; i < shape.N; i++)
+            {
+            shape.x[i] += (OverlapReal)dr.x;
+            shape.x[i] *= scale;
+            shape.y[i] += (OverlapReal)dr.y;
+            shape.y[i] *= scale;
+            shape.z[i] += (OverlapReal)dr.z;
+            shape.z[i] *= scale;
+            vec3<Scalar> vert(shape.x[i], shape.y[i], shape.z[i]);
+            rsq = fmax(rsq, dot(vert, vert));
+            }
+        shape.diameter = OverlapReal(2.0 * fast::sqrt(rsq));
+        }
+
+    Scalar getVertexMoveProbability()
+        {
+        return this->m_move_probability;
+        }
+
+    void setVertexMoveProbability(Scalar vertex_move_prob)
+        {
+        this->m_move_probability = fmin(vertex_move_prob, 1.0);
+        }
+
+    void prepare(uint64_t timestep) { }
+
+    void update_shape(uint64_t timestep,
+                      Scalar& stepsize,
+                      const unsigned int& type_id,
+                      param_type& shape,
+                      hoomd::RandomGenerator& rng)
+        {
+        if (!m_calculated[type_id])
+            {
+            detail::MassProperties<ShapeConvexPolyhedron> mp(shape);
+            this->m_centroids[type_id] = mp.getCenterOfMass();
+            m_calculated[type_id] = true;
+            }
+        // mix the shape.
+        for (unsigned int i = 0; i < shape.N; i++)
+            {
+            if (hoomd::detail::generate_canonical<double>(rng) < this->m_move_probability)
+                {
+                vec3<Scalar> vert(shape.x[i], shape.y[i], shape.z[i]);
+                move_translate(vert, rng, stepsize, 3);
+                shape.x[i] = (OverlapReal)vert.x;
+                shape.y[i] = (OverlapReal)vert.y;
+                shape.z[i] = (OverlapReal)vert.z;
+                }
+            }
+        detail::MassProperties<ShapeConvexPolyhedron> mp(shape);
+        Scalar volume = mp.getVolume();
+        vec3<Scalar> dr = this->m_centroids[type_id] - mp.getCenterOfMass();
+        OverlapReal scale = (OverlapReal)fast::pow(this->m_volume[type_id] / volume, 1.0 / 3.0);
+        scaleParticleVolume(shape, dr, scale);
+        stepsize *= scale;
+        }
+
+    void retreat(uint64_t timestep)
+        {
+        // move has been rejected.
+        }
+
+    private:
+    std::vector<bool> m_calculated; // whether or not mass properties has been calculated
+    };                              // end class ConvexPolyhedronVertexShapeMove
+
+template<class Shape> class ElasticShapeMove : public ShapeMoveBase<Shape>
```

Comment:
The `ConvexPolyhedron` specific logic should go in the template specification. In the Ellipsoid shape, you overwrite most of the code.